internal/plan/plan.go

```diff
@@ -184,6 +187,76 @@ func (s *Service) Equal(other *Service) bool {
 	return reflect.DeepEqual(s, other)
 }
 
+// GetCommand overrides the default arguments (inside [ ]) if
+// CmdArgs is present and returns the command as a stream of strings.
+func (s *Service) GetCommand() ([]string, error) {
+	args, err := shlex.Split(s.Command)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse service %q command: %s", s.Name, err)
+	}
+	fargs := make([]string, 0)
+	for _, arg := range args {
+		if arg == "[" || arg == "]" {
+			if len(s.CmdArgs) > 0 {
+				break
+			} else {
+				continue
+			}
+		}
+		fargs = append(fargs, arg)
+	}
+	for _, arg := range s.CmdArgs {
+		fargs = append(fargs, arg)
+	}
+	return fargs, nil
+}
+
+// GetCommandStr takes in the output of GetCommand and
+// returns the command as a string in good print format
+func (s *Service) GetCommandStr(cmd []string) string {
+	f := func(str string) string {
+		for _, c := range str {
+			if c < '!' {
+				return strconv.Quote(str)
+			}
+		}
+		return str
+	}
+	args := make([]string, 0)
+	for _, arg := range cmd {
+		args = append(args, f(arg))
+	}
+	return strings.Join(args, " ")
+}
+
+func (s *Service) setArgs(args []string) error {
+	s.CmdArgs = append([]string(nil), args...)
+	return nil
+}
+
+func (s *Service) checkCommand() error {
```

Comment:
Ah, I see you're checking some of the error cases here. However, it looks like this checks that the number of `[` and `]` is balanced (and equal to 1) but not that the `[` comes before the `]`. I think my code linked above handles that.

It is good to validate command here in the plan parsing code -- thanks. Perhaps we can share and reuse a function like `GetCommand()` here and in the service manager, and just pass `nil` to it for `cmdArgs` here when validating.

Reply:
Yes, your code does indeed handle it. 
Mine does as well, but I guess, quite subtly.
```
	if leftCnt > 0 || rightCnt > 0 {
		if leftCnt != 1 || rightCnt != 1 || args[len(args)-1] != "]" {
			return fmt.Errorf("bad syntax regarding optional/overridable arguments")
		}
	}
```
Here, if `]` somehow comes before `[` and the count of them is still 1, the `args[len(args)-1] != "]"` check will fail. It's not quite transparent nor intuitive, sorry about that. Tried to be a bit cheeky.
I did plan on writing something like your proposed function, but couldn't resist to be a bit cheeky. :stuck_out_tongue: 
I will write it up nicely, using references from your code, thanks!

Reply:
Added, thanks!